File: 3200/main.go
package main

import "fmt"

func main() {
	fmt.Println(maxHeightOfTriangle(1, 1))	// 1
	fmt.Println(maxHeightOfTriangle(2, 4))	// 3
	fmt.Println(maxHeightOfTriangle(2, 1))	// 2
	fmt.Println(maxHeightOfTriangle(10, 1))	// 2

}

func maxHeightOfTriangle(red int, blue int) int {
	return max(f(red, blue), f(blue, red))
}

func f(red, blue int) int {
	for i := 1; ; i++ {
		if i % 2 == 1 {
			red -= i
			if red < 0 {
				return i - 1
			}
		} else {
			blue -= i
			if blue < 0 {
				return i - 1
			}
		}
	}
}

File: 3200/main.rs
struct Solution {}

impl Solution {
    pub fn max_height_of_triangle(red: i32, blue: i32) -> i32 {
        fn h(x: i32, y: i32) -> i32 {
            let odd = 2 * ((x as f64).sqrt() as i32) - 1;
            let even = 2 * (((-1.0 + (1.0 + 4.0 * (y as f64)).sqrt())) / 2.0) as i32;
            odd.min(even) + 1
        }
        std::cmp::max(h(red, blue), h(blue, red))
    }
}

fn main() {
    println!("{}", Solution::max_height_of_triangle(10, 1));
}